### ISSUE TITLE
incorrect NodeIDAllocator behavior with nodes added immediately after boot

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#3262]
run these lines in order, waiting a few seconds in between:

    ServerTree.add { Synth(\default) };
    s.boot;
    Synth(\default);

the second synth fails with

    FAILURE IN SERVER /s_new duplicate node ID

but instantiating further synths succeeds.

this is a distillation of an issue i've been having with the SafetyNet quark that appeared only recently. i think it may be a side effect of #3245.

**Comment by telephon** [comment]
Maybe it is here:

`bootInit` calls 
`if(recover) { this.newNodeAllocators } { this.newAllocators };`

and then later `this.initTree;`, which in turn calls:

`this.newNodeAllocators;`

This is certainly wrong.

**Comment by telephon** [comment]
I don't think that `initTree`should call `this.newNodeAllocators` at all.

**Comment by telephon** [comment]
Actually, `newNodeAllocators` is called by `clientID_` and this happens after `initTree`. The whole startup sequence is really opaque, we should fix that.

**Comment by adcxyz** [comment]
> I don't think that initTreeshould call this.newNodeAllocators at all. Any opinions? @muellmusik ? @adcxyz ?

I left this as was as much as possible.
 
> The whole startup sequence is really opaque, we should fix that.

Yes! Very happy to help with that, not much time today, tied up in two symposia at the same time ;-)

> It seems that clientID_ should trigger the full initialization.

Yes. Briefly, I think the proper boot sequence should be:
boot scsynth and send nofify request
// currently, scsynth is inited already with its given clientID or 0
// maybe we should even lock the server until answer arrives
on scsynth response, init all allocators (because now we have a valid clientID)
then run all functions registered to do on boot. 

No time to look it up now, but it seems the synth made from ServerTree func
got its nodeID from the initial allocator, and the allocator newly made  after the 
scsynth response starts over with the same initial node id. 

so if ServerTree happens after the response, it should be fine.
more later, 
adc

**Comment by telephon** [comment]
what should happen for any running synths when you set the clientID?

**Comment by muellmusik** [comment]
> I don't think that initTreeshould call this.newNodeAllocators at all. Any opinions? @muellmusik ? @adcxyz ?

I don’t off the top of my head see why it should need to. Encapsulation says it shouldn’t matter what the ids are as that’s an internal problem, and of course the point of the allocators is to hide that detail from the user, so it doesn’t necessarily need a reset.

The devil will be in the details but this could simplify things.

**Comment by jamshark70** [comment]
> The whole startup sequence is really opaque, we should fix that.

It really needs a proper spec where we think through not only the ideal sequence, but also what should happen when users do things that *don't* make sense, e.g.:

1. Call `boot` 2 or 3 or 10 times. In Windows particularly, I've sometimes seen the boot process to be quite slow to respond, which may invite impatient users to try again. We've defined the behavior if you `boot` once and wait for everything to proceed, and we've defined the behavior if you `boot` again after the server is fully booted. I'm not sure of the behavior if you `boot` while there's another `boot` call pending. Quite possibly, it's OK in the current version -- but maybe purely by luck.

2. Call `boot` and then `quit` before the server is fully booted. Do we cancel pending functions? I guess we don't cancel `ServerBoot` functions, but `waitForBoot`? I did a little bit of testing and got some indeterminate behavior. In one test run (code below), scsynth started and quit *without* doing the waitForBoot, then it started again and did *both* waitForBoot actions -- that is, a sequence of waitForBoot - quit - waitForBoot is not guaranteed to leave the server in the same state that it would be in if you only did the second waitForBoot.

3. Others?

It may seem to be pedantic, but one of the reasons why the boot process is so blastedly complicated is that users found they could get into weird inconsistent states, and we fixed those by adding hacks upon hacks upon hacks, to the point that `mBootNotifyFirst` is in there but we don't know what it means anymore (to cite one question that is still pending).

Good software design takes both correct and incorrect input into account.

> what should happen for any running synths when you set the clientID?

Good question. Of course, users "shouldn't," but it will happen...

```
// boot-quit-boot test
(
var respFuncWithId = { |id| { "Reponse %\n".postf(id) } };

fork {
	s.quit;
	5.wait;
	
	s.waitForBoot(respFuncWithId.(UniqueID.next));
	0.5.wait;
	s.quit;
	5.wait;
	s.waitForBoot(respFuncWithId.(UniqueID.next));
};
)
```

**Comment by adcxyz** [comment]
@jamshark70 couldn't agree more, 
lets write a full spec first of all cases to cover, 
and what exactly should happen in which  circumstances. 

For the bug reported here, I would propose to remove initTree from bootInit : 
```
	bootInit { | recover = false |
		if(recover) { this.newNodeAllocators } { this.newAllocators };
		if(dumpMode != 0) { this.sendMsg(\dumpOSC, dumpMode) };
		if(sendQuit.isNil) {
			sendQuit = this.inProcess or: { this.isLocal };
		};
		this.connectSharedMemory;
		// this.initTree;  // <- too early here, do after scsynth response
	}
```
best adc

**Comment by adcxyz** [comment]
BTW, this topic branch tried to cover some of the more annoying cases, 
so maybe some of it can be recycled:
https://github.com/adcxyz/supercollider/tree/topic-server-boot-recover

**Comment by jamshark70** [comment]
One thing I noted in the other "untidy boot behavior" issue: IMO the ideal way to handle user boot actions (functions passed to waitForBoot or doWhenBooted) is to save them in a queue.

- Start one and only one "waiting for boot" routine. Multiple waitForBoot or doWhenBooted calls should check for the existence of a routine and *not* create extra ones. Simple reason: Syncing multiple threads is hard.
- Functions passed to waitForBoot or doWhenBooted only get added into the queue, nothing else.
- `quit`, or boot failure, should clear the queue.
- The single "waiting for boot" thread then does this, after the server process is up:
  1. Shared memory interface
  2. initTree
  3. ServerBoot
  4. User queue.

Clean, streamlined, leaving nothing to the scheduler's whims. We would be certain that user boot actions evaluate in the order in which the user added them.

That's pretty extensive surgery in the boot process, but I believe it would eliminate an entire category of boot-sequence problems, and simplify future additions. IMO we're wasting time if we do anything less.

**Comment by muellmusik** [comment]
That sounds sensible to me, James. Though I'm not totally sure about the order. There may be more steps than really needed. What if whatForBoot doWhenBooted actually just become shortcuts for booting plus a call to ServerBoot:doOnce. (The latter doesn't exist, but could)

For users I think a little more clarity in the doc as to the distinctions between and use cases for ServerTree and ServerBoot would be helpful.

**Comment by adcxyz** [comment]
> >  @telephon:
> >     I don't think that initTreeshould call this.newNodeAllocators at all. Any opinions? 
> @muellmusik :
> I don’t off the top of my head see why it should need to. Encapsulation says it shouldn’t matter what the ids are as that’s an internal problem, and of course the point of the allocators is to hide that detail from the user, so it doesn’t necessarily need a reset.

I agree, the reset is not strictly necessary - it only needs to be remade when clientID changes. Actually, doing newNodeAllocators in initTree also throws away permanent nodeIDs when doing CmdPeriod, which does not seem like a good idea at all.

So, is there agreement to change initTree like this?
```
	initTree {
-		this.newNodeAllocators;
		this.sendDefaultGroups; 
		tree.value(this);
		ServerTree.run(this);
	}
/// tests:
~perms = 10.collect { s.nextPermNodeID }; // (2..11)
Pbind(\degree, Pseq((0..7)), \dur, 0.1).play; // 1000, ... 1009

CmdPeriod.run;

().play; // 1010, count continues
s.nextPermNodeID; // 12, perms continue

s.reboot;
// new allocator made:
s.nextPermNodeID; // 2, 
s.nextNodeID; // 1000
```

**Comment by adcxyz** [comment]
hi @snappizz, 
Since you mentioned using the Safety quark, I tested how well that runs with 3.9, 
and found several things to improve. If you like, please have a look: 
https://github.com/adcxyz/SafetyNet/commit/d9a0fa5

thanks adc

**Comment by muellmusik** [comment]
> I agree, the reset is not strictly necessary - it only needs to be remade when clientID changes. Actually, doing newNodeAllocators in initTree also throws away permanent nodeIDs when doing CmdPeriod, which does not seem like a good idea at all.
> 
> So, is there agreement to change initTree like this?

See my comment here: https://github.com/supercollider/supercollider/pull/3265#discussion_r147819080

I think we can remove resetting entirely with the 'Readable' approach. As far as permanent IDs go, I can't say I've ever really understood the point as they seem to violate encapsulation, i.e. it seems to me either you're hard coding IDs, or you're encapsulating them behind the allocation scheme, in which case why do you need them anyway? IDs are the allocator's problem, not yours.

That said I don't want to hose anybody's implementation by removing them, and I think it's no problem anyway. With the readable approach there's probably no real need to distinguish between permanent and non-permanent IDs in most usage. If the ID is clientID-32bitInteger it's unlikely to wrap around anyway, but you could hold permanent ones in a reserved pool if needed. Or maybe simpler just use clientID-perm-32bitInteger and a separate allocator for permanent ones.

**Comment by jamshark70** [comment]
By not resetting, effectively all node IDs are permanent within the scope of an sclang session ;)

On the permanent ID issue, I give up. The scenario was MixerChannels that need to rebuild on CmdPeriod while other objects need access to the mixer's groups as synth targets. The mixer has a `synthgroup` and an `effectgroup`, so `asTarget` will not cover all cases. Therefore the client object (such as a Voicer) may need to hold a reference to the group object itself. So the only way for the mixer to rebuild, without permanent IDs, is to rigorously retain the identical group objects, get new IDs from allocator (because, if it doesn't, there's no guarantee of being in sync with the allocator), and generate e.g. `fadergroup.newMsg`.

**Comment by jamshark70** [comment]
Sorry about that. Github needs to remove that "close and comment" button from the mobile site. Don't they know touch screens are not as precise as a mouse? No confirmation? "Oops" but you can't recall the action.

Anyway. I'm ok with that if this is really the best solution.

You will freak users out if they reboot and get ID 189274 for the next synth, though: Be prepared to explain about encapsulation a few hundred times on the list ;)

**Comment by muellmusik** [comment]
> By not resetting, effectively all node IDs are permanent within the scope of an sclang session ;)

Yeah, well this was sort of my point. We're making it harder than it needs to be by adding an extra requirement.



> You will freak users out if they reboot and get ID 189274 for the next synth, though: Be prepared to explain about encapsulation a few hundred times on the list ;)

Well, I'm happy to field those. Of course the IDs don't need to be numerical. For super robustness we could use UUIDs. We should consider what's useful. User supplied labels are also possible and would be handy for many reasons. ClientID-filtSynth-7826.

**Comment by vivid-synth** [comment]
> Of course the IDs don't need to be numerical. For super robustness we could use UUIDs. We should consider what's useful. User supplied labels are also possible and would be handy for many reasons. ClientID-filtSynth-7826.

Are we talking about in sclang or on the server?

**Comment by snappizz** [comment]
reopening because the fix is not yet in 3.9.

**Comment by brianlheim** [comment]
Closed by #3273.
